### PR TITLE
Fix issues regarding cloneWithType() with int subtypes

### LINKED ISSUE #34082: `cloneWithType` should not panic in some cases with int subtypes
**Description:**

In the below cases `cloneWithType` should not panic because values 1, 2, 3 belong to all the int subtypes and hence `cloneWithType` does not require any conversion. So it should pass and should not panic.
```bal
public function main() {
    int[] x = [1, 2, 3];
    (int|decimal)[] e1 = checkpanic x.cloneWithType(); // does not panic, correct behaviour
    (byte|float)[] e2 = checkpanic x.cloneWithType(); // panics, but should not 
    (int:Unsigned8|float)[] e3 = checkpanic x.cloneWithType(); // panics, but should not 
    (int:Signed8|decimal)[] e4 = checkpanic x.cloneWithType(); // panics, but should not 
    (decimal|int:Signed16)[] e5 = checkpanic x.cloneWithType(); // panics, but should not 

    byte[] y = [1,2,3];
    (byte|float)[] p1 = checkpanic y.cloneWithType(); // does not panic, correct behaviour
    (int|float)[] p2 = checkpanic y.cloneWithType(); // panics, but should not 
    (int:Unsigned8|float)[] p3 = checkpanic y.cloneWithType(); // panics, but should not 

    int:Signed16[] z = [1, 2, 3];
    (int|float)[] q1 = checkpanic z.cloneWithType(); // does not panic, correct behaviour
    (byte|float)[] q2 = checkpanic z.cloneWithType(); // panics, but should not
    (int:Signed32|float)[] q3 = checkpanic z.cloneWithType(); // panics, but should not
}
```
Should also consider the other remaining cases of int subtypes not shown above.

**Steps to reproduce:**

**Affected Versions:**
Ballerina Swan Lake Beta 5 - snapshot

**OS, DB, other environment details and versions:**

**Related Issues (optional):**

**Suggested Labels (optional):**

**Suggested Assignees (optional):**

### PR DESCRIPTION
**Pull request by gabilang**
Fixes #34082

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/TypeConverter.java [hunk 1]
```diff
@@ -506,6 +507,12 @@ private static void addErrorMessage(int errorCountDifference, List<String> error
         }
     }
 
+    private static boolean integerSubtypeMembers(Type inputValueType, Type targetType) {
```
Comment (HindujaB) [inline bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/TypeConverter.java:510]: Shall we rename this method something like `isIntegerSubtypeMember()` or `checkIntegerSubtypeMember()` ? It is preferred the method name to be a verb.

File: bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/TypeConverter.java [hunk 2]
```diff
@@ -274,7 +274,8 @@ static boolean isConvertibleToFloatingPointTypes(Object value) {
         switch (targetTypeTag) {
             case TypeTags.UNION_TAG:
                 for (Type memType : ((BUnionType) targetType).getMemberTypes()) {
-                    if (TypeChecker.getType(inputValue) == memType) {
+                    Type inputValueType = TypeChecker.getType(inputValue);
+                    if (inputValueType == memType || checkIntegerSubtypeMembers(inputValueType, memType)) {
```
Comment (Nadeeshan96) [inline bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/TypeConverter.java:278]: It seems currently with this fix it gives incorrect behaviour for some cases. For example
```bal
public function main() {
    int[] z = [257, 257, 257];
    (byte|int:Unsigned16)[] l = checkpanic z.cloneWithType();
}
```
previously succeeded in conversion as it should be. Now it fails giving the below result.
```logtalk
Running executable

error: {ballerina}NumberConversionError {"message":"'int' value '257' cannot be converted to 'byte'"}
        at ballerina.lang.value.0:cloneWithType(value.bal:86)
           temp:main(temp.bal:5)
```

Could you please add more tests covering different ranges in int subtypes.

Comment (gabilang) [reply bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/TypeConverter.java]: Resolved with https://github.com/ballerina-platform/ballerina-lang/pull/34136/commits/980da37aaf38e20954dceb368edf867887ae3ce1

File: bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/TypeConverter.java [hunk 3]
```diff
@@ -274,7 +274,9 @@ static boolean isConvertibleToFloatingPointTypes(Object value) {
         switch (targetTypeTag) {
             case TypeTags.UNION_TAG:
                 for (Type memType : ((BUnionType) targetType).getMemberTypes()) {
-                    if (TypeChecker.getType(inputValue) == memType) {
+                    if (TypeChecker.getType(inputValue) == memType
+                            || isIntegerSubtypeAndConvertible(inputValue, memType)
+                    ) {
```
Comment (warunalakshitha) [inline bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/TypeConverter.java:279]: Should this closing parenthesis need goto line 278. seems format error

Comment (gabilang) [reply bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/TypeConverter.java]: Addressed with https://github.com/ballerina-platform/ballerina-lang/pull/34136/commits/0af5ad271764dfb727df0e90bda5905cbb4ed362

File: bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/TypeConverter.java [hunk 4]
```diff
@@ -606,33 +629,51 @@ private static boolean isConvertibleToTupleType(Object sourceValue, BTupleType t
...
-                }
+                returnVal = false;
+            } else if (convertibleTypes.size() > 1 && !convertibleTypes.contains(TypeChecker.getType(
+                    source.getRefValue(i))) && !hasIntegerSubTypes(convertibleTypes)) {
+                addErrorMessage(errors.size() - initialErrorCount, errors,
+                        "tuple element '" + elementIndex + "' cannot be converted to '" +
+                                targetTypes.get(i) + "': ambiguous target type");
                 returnVal = false;
             }
+            if (errors.size() >= MAX_CONVERSION_ERROR_COUNT + 1) {
```
Comment (warunalakshitha) [inline bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/TypeConverter.java:650]: Is this cover in the tests? There are a few more lines showed as not covered in tests

Comment (gabilang) [reply bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/TypeConverter.java]: Removed unnecessary checks and added a test case for it.

**Review by warunalakshitha — APPROVED** [review]
LGTM

**Pull request merged**